### PR TITLE
Concurrency for systems

### PR DESCRIPTION
**Pull request by ianling**
Contains all the changes to make Systems all run in parallel with each other in their own goroutines.

Each system has a target "tick rate", which is how many times it wants to "tick" (run `system.Update()`) every second.

When you add a system to the world (`w.AddSystem(&mySystem{})`), the system is initialized, and then put in a queue to be started on the next world update. When the system is started, it begins updating at its target tick rate in its own goroutine.

Mainly for testing purposes, you can also choose not to start a system and single-step it at your own pace instead.

Each system keeps track of its own `lastTick` timestamp, since every system ticks independently.

Also contains some random thread-safety fixes.

### CODE REVIEW

File: system.go [hunk 1]
```diff
@@ -1,32 +1,28 @@
...
-	System
-	bind(*World)
-}
-
-// SystemInitializer is a system with an Init method
-type SystemInitializer interface {
-	System
-	Init(*World)
-	IsInitialized() bool
-}
```
Comment (gravestench) [inline system.go:22]: This interface was broken out to allow for systems that don't have any initialization logic, basically just to reduce `noop` methods an end user might need to make when creating a new system.

File: world.go [hunk 2]
```diff
@@ -16,13 +16,14 @@ type componentFactories = map[ComponentID]*ComponentFactory
 func NewWorld(cfg *WorldConfig) *World {
 	world := &World{
 		entityManagement: &entityManagement{
-			ComponentFlags: make(map[EID]*bitset.BitSet),
-			Subscriptions:  make([]*Subscription, 0),
+			nextEntityID:       new(uint64),
+			Subscriptions:      make([]*Subscription, 0),
 			entityRemovalQueue: make([]EID, 0),
```
Comment (gravestench) [inline world.go:21]: Interesting. Why store the `nextEntityID` as a pointer to a uint64 ?

File: world.go [hunk 3]
```diff
@@ -42,29 +43,38 @@ func NewWorld(cfg *WorldConfig) *World {
...
 
 type systemManagement struct {
 	Systems            []System
+	systemStartQueue   []func()
 	systemRemovalQueue []System
 }
 
 // World contains all of the Entities, Components, and Systems
 type World struct {
+	// TimeDelta is the time since the last tick occurred
```
Comment (gravestench) [inline world.go:66]: It's about time this was documented!

File: world.go [hunk 4]
```diff
@@ -79,9 +89,8 @@ func (w *World) RegisterComponent(c Component) ComponentID {
 		return id
 	}
 
-	nextID := ComponentID(len(w.factories))
-
-	factory := newComponentFactory(nextID)
+	nextId := atomic.AddUint64(w.nextEntityID, 1)
```
Comment (gravestench) [inline world.go:92]: is this a concurrency thing?

File: world.go [hunk 5]
```diff
@@ -296,7 +339,12 @@ func (w *World) updateComponentFlags(id EID) {
 
 		// ... for the bit index that corresponds to the map id,
 		// use the true/false value we just obtained as the value for that bit in the bitset.
-		w.ComponentFlags[id].Set(int(w.factories[idx].ID()), found)
+		cf, cfFound := w.ComponentFlags.Load(id)
+		if !cfFound {
+			continue
+		}
```
Comment (gravestench) [inline world.go:345]: We almost certainly should explode if the component flags aren't found. We need to write more tests that hit this and really think about what it means for us to wind up here, and what we should do about it.

File: world.go (hunk 2 above)
Comment (ianling) [reply world.go:21]: The `atomic` stuff (basically thread-safe access to uints/ints) operates on pointers, and to keep this threadsafe, we should always access this number through `atomic`, so it just made more sense to me to store this as a pointer to make people think twice about dereferencing it/remind them that this needs to be accessed via `atomic`

File: world.go (hunk 4 above)
Comment (ianling) [reply world.go:92]: Yeah, `atomic` is a thread-safe thing that allows you to add/subtract and read a uint/int from different threads without needing to explicitly guard it with a mutex

File: system.go [hunk 6]
```diff
@@ -1,32 +1,28 @@
...
-
-type SystemUpdater interface {
-	System
 	Update()
+
+	Tick()
```
[the remark this answers is not in the document]
Comment (ianling) [reply system.go:12]: Oh, my bad, I was mixing up the commit you can see here with my unpushed changes. It's pretty unclear in this commit what the difference is. We should take another look at this once I've pushed my unpushed changes

[the remark this answers is not in the document]
Comment (gravestench) [reply system.go:12]: All good, I was just asking as I was reading through the changes.

This is good stuff.

File: system.go (hunk 1 above)
Comment (ianling) [reply system.go]: For the time being, I ended up reverting back to something like this SystemInitializer thing

File: world.go [hunk 7]
```diff
@@ -42,29 +43,38 @@ func NewWorld(cfg *WorldConfig) *World {
...
 	*entityManagement
 	*componentManagement
 	*systemManagement
+	// mutex locks access to various World resources to maintain thread safety.
+	// This should be locked when accessing any shared World resources, like slices and maps
 	mutex sync.Mutex
+	// tickWaitgroup tracks how many systems are still processing the current tick,
+	// allowing World to wait for them to finish before moving on to the next tick
+	tickWaitgroup  sync.WaitGroup
+	targetTickRate float32
```
[the remark this answers is not in the document]
Comment (ianling) [reply world.go:77]: Woops, those are actually unused leftovers from a previous iteration of this, I'm gonna remove them entirely

**Pull request merged**